designs/PackageVulnerability/FlagVulnerablePackages.md

```diff
@@ -0,0 +1,142 @@
+
+#### ✔️ In scope (Stage 1)
+
+* For CLI:
+  * A CLI command to **list vulnerabilities** in a project/solution project graph
+
+* For Visual Studio:
+  * **Display vulnerability information** in the package graph on `Installed` tab of Package Manager Dialog (PMUI)
+  * **Show transitive packages for `PackageReference` type projects** in PMUI
+
+* For nuget.org:
+  * **Automatically ingest and refresh NuGet package vulnerability data from GitHub**
+  * **Surface package vulnerability information on-demand**:
+    * as part of package metadata in `v3` registration blobs (`JSON`)
+    * by showing vulnerability information to package owners (only!) on package details page and version history
+  * Basic (e-mail) notification to notify package publishers for any new vulnerability detected in one or more of their packages
+
+#### ❌ Out of scope (Stage 1)
+
+Some enhancements to this feature are out of scope for the initial implementation of this feature because there are several unknowns still to be clarified. See also the *Open Questions* section below.
+
+* Enable `restore` to Inform or Warn (see Stage 2)
+* Expose vulnerability metadata in nuget.org *Search* service (not required for MVP)
+* Display vulnerability information for *transitive packages* in PMUI (additional UX complexity)
+* Expose vulnerability information *publicly* on nuget.org's package details page and version history (disclosure policies TBD)
+
+### Stage 2
+
+Stage 2 of this feature will focus on *highlighting known vulnerabilities when the package is already installed*. This involves introducing new validations at package `restore` time, and new ways of dealing with them.
+
+#### ✔️ In scope (Stage 2)
+
+* For CLI:
+  * Output warnings or errors when running `dotnet add package` (which actually does a *restore*)
+
+* For Visual Studio:
+  * Inform consumers about known vulnerabilities at time of installation or restore:
+    * Warn (⚠️) for `critical` and `high` impact vulnerabilities (minimal info + command/way to uncover detailed information)
+    * Info (ℹ️) for lower impact vulnerabilities (minimal info + command/way to uncover detailed information)
+  * Display vulnerability information for *transitive packages* in PMUI
+  * New settings to enable/disable warnings for the various impact levels of vulnerabilities
+
+* For nuget.org:
+  * Show vulnerability information in packages to their publishers on nuget.org
+    * Show the vulnerability info on the package page and version history
+
+#### ❓ Open questions (Stage 2)
+
+* For `restore`, should it be opt-in or opt-out (preferred for security features)?
+* For `dotnet add package`, define behavior (info, warn, error?) when encountering packages known to be vulnerable.
```

Comment:
None. We don't do this for other audits & generally we are just advising people on restore to not use these packages & providing them a resolution. Everything should be in dotnet restore instead.

Reply:
I'll adjust the spec to reflect this, but I want to be sure that we're on the same page. `dotnet add package` includes a `dotnet restore` unless the `--no-restore` option is included (https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-add-package). So, we will have `dotnet restore` (above) defaulting to indicating vulnerabilities, and `dotnet add package` will have the same behavior as `dotnet restore` except when `--no-restore` is used. Right?